### PR TITLE
Fix SignalConnector._has_already_handler check for callable type

### PR DESCRIPTION
**Pull request by mauvilsa**
## What does this PR do?

Fixes an issue where a user's registered SIGTERM signal handler is overwritten by `SignalConnector.register_signal_handlers`. One case in which this happens is when a method of a class instance is registered. That is:

```python
import signal

class Callback:
    def __init__(self):
        signal.signal(signal.SIGTERM, self.sigterm_handler)
    def sigterm_handler(self):
        pass

callback = Callback()
```

The current check `isinstance(signal.getsignal(signal.SIGTERM), FunctionType)` evaluates to false, thus overwriting it.

No related issue. This was discussed over slack.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by carmocca — COMMENTED** [review]
LGTM. Can you add a test to avoid regressions? Can be a simple unit-test. `_has_already_handler` can be made static for it.

**Comment by mauvilsa** [comment]
> Can you add a test to avoid regressions? Can be a simple unit-test.

Done.

File: tests/trainer/connectors/test_signal_connector.py [hunk 1]
```diff
@@ -57,3 +58,27 @@ def training_step(self, batch, batch_idx):
         else:
             trainer.fit(model)
         assert trainer._terminate_gracefully == (False if register_handler else terminate_gracefully)
+
+
+def signal_handler():
+    pass
+
+
+class C:
```
Comment (awaelchli) [inline tests/trainer/connectors/test_signal_connector.py:67]:
```suggestion
class SignalHandlers:
```

File: tests/trainer/connectors/test_signal_connector.py [hunk 2]
```diff
@@ -57,3 +58,27 @@ def training_step(self, batch, batch_idx):
...
+@pytest.mark.parametrize(
+    ["handler", "expected_return"],
+    [
+        (signal.Handlers.SIG_DFL, False),
+        (signal_handler, True),
+        (C().signal_handler, True),
+    ],
+)
+def test_has_already_handler(handler, expected_return):
+    trainer = Trainer()
```
Comment (awaelchli) [inline tests/trainer/connectors/test_signal_connector.py:81]:
```suggestion
def test_has_already_handler(handler, expected_return):
    """Test that the SignalConnector detects whether a signal handler is already attached."""
    trainer = Trainer()
```

**Review by awaelchli — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/trainer/connectors/signal_connector.py [hunk 3]
```diff
@@ -94,7 +94,7 @@ def _is_on_windows(self) -> bool:
 
     def _has_already_handler(self, signum: Signals) -> bool:
         try:
-            return isinstance(signal.getsignal(signum), FunctionType)
+            return callable(signal.getsignal(signum))
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/signal_connector.py:97]: Seeing test failures here

```python
_________________ test_fault_tolerant_sig_handler[True-False] __________________

register_handler = False, terminate_gracefully = True
tmpdir = local('/tmp/pytest-of-root/pytest-0/test_fault_tolerant_sig_handle0')

    @pytest.mark.parametrize("register_handler", [False, True])
    @pytest.mark.parametrize("terminate_gracefully", [False, True])
    @RunIf(skip_windows=True)
    def test_fault_tolerant_sig_handler(register_handler, terminate_gracefully, tmpdir):
    
        if register_handler:
    
            def handler(*_):
                pass
    
            signal.signal(signal.SIGUSR1, handler)
    
        class TestModel(BoringModel):
            def training_step(self, batch, batch_idx):
                if terminate_gracefully or register_handler:
                    os.kill(os.getpid(), signal.SIGUSR1)
                    sleep(0.1)
                return super().training_step(batch, batch_idx)
    
        model = TestModel()
    
        with mock.patch.dict(os.environ, {"PL_FAULT_TOLERANT_TRAINING": str(int(terminate_gracefully))}):
    
            trainer = Trainer(default_root_dir=tmpdir, max_epochs=1, limit_train_batches=2, limit_val_batches=0)
            if terminate_gracefully and not register_handler:
                with pytest.raises(ExitGracefullyException):
>                   trainer.fit(model)
E                   Failed: DID NOT RAISE <class 'pytorch_lightning.utilities.exceptions.ExitGracefullyException'>

/__w/pytorch-lightning/pytorch-lightning/tests/trainer/connectors/test_signal_connector.py:56: Failed
```

**Comment by awaelchli** [comment]
Interesting, wasn't aware of this PR, but I have also one here #10611 where I changed the `has_already_sampler` to return

`return signal.getsignal(signum) is not signal.SIG_DFL`

Is the check for callable necessary? Which way is the correct way to check this, anyone knows? :)

File: pytorch_lightning/trainer/connectors/signal_connector.py (hunk 3 above)
Comment (mauvilsa) [reply pytorch_lightning/trainer/connectors/signal_connector.py:97]: I don't know why this is failing. Locally all tests in `test_signal_connector.py` are run and are successful. And the initial runs for this pull request were successful, see e.g. https://github.com/PyTorchLightning/pytorch-lightning/actions/runs/1451925254. So it must be related to some other pull request that was merged recently.

**Comment by mauvilsa** [comment]
> Is the check for callable necessary? Which way is the correct way to check this, anyone knows? :)

I guess the callable is not necessary. What should be the behavior if the value is `signal.SIG_IGN` which is part of the same enum that defines `signal.SIG_DFL`.

**Comment by awaelchli** [comment]
Good point, I think we should respect it and not add a handler in that case too?

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/signal_connector.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/signal_connector.py]: Hey, I added a few commits here but couldn't solve the issue. But I believe it could be related to the missing teardown I have seen over here #10611. In fact, if I include your change in signal connector, the tests pass there. So I suggest we merge my teardown fix first, then attempt to run tests again here.
